### PR TITLE
sbt new and Templates

### PR DESCRIPTION
**Pull request by eed3si9n**
/review @dwijnand

### CODE REVIEW

File: src/reference/02-DetailTopics/05-Plugins-and-Best-Practices/06-sbt-new-and-Templates.md [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+```
+\$ sbt new foundweekends/giter8.g8
+```
+
+#### Use CC0 1.0 for template licensing
+
+We recommend licensing software templates under [CC0 1.0][CC0],
+which waives all copyrights and related rights, similar to the "public domain."
+
+If you reside in a country covered by Berne Convention, such as the US,
```
Comment (dwijnand) [inline src/reference/02-DetailTopics/05-Plugins-and-Best-Practices/06-sbt-new-and-Templates.md]: by the Berne Convention

File: src/reference/02-DetailTopics/05-Plugins-and-Best-Practices/06-sbt-new-and-Templates.md [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+copyright will arise automatically without registration.
+Thus, people won't have legal right to use your template if you do not
+declare the terms of license.
+The tricky thing is that even the most permissive license such as MIT License and Apache License will require attribution in the software.
```
Comment (dwijnand) [inline src/reference/02-DetailTopics/05-Plugins-and-Best-Practices/06-sbt-new-and-Templates.md]: ... even permissive licenses such as the MIT license and the Apache 2.0 License require attribution to the software.

File: src/reference/02-DetailTopics/05-Plugins-and-Best-Practices/06-sbt-new-and-Templates.md [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+To remove all claims to the templated snippets, distribute it under CC0, which is an international equivalent to public domain.
+
+```
+License
+-------
+Copyright 20XX, Foo, Inc.
+This template is distributed under [CC0 1.0](https://creativecommons.org/publicdomain/zero/1.0/), equivalent of public domain.
```
Comment (dwijnand) [inline src/reference/02-DetailTopics/05-Plugins-and-Best-Practices/06-sbt-new-and-Templates.md]: We should copy and/or link to the official way to apply CC0: https://wiki.creativecommons.org/wiki/CC0_FAQ#May_I_apply_CC0_to_computer_software.3F_If_so.2C_is_there_a_recommended_implementation.3F

File: src/reference/02-DetailTopics/05-Plugins-and-Best-Practices/06-sbt-new-and-Templates.md [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+```
+
+### How to extend sbt new
+
+The rest of this page explains how to extend `sbt new` command
```
Comment (dwijnand) [inline src/reference/02-DetailTopics/05-Plugins-and-Best-Practices/06-sbt-new-and-Templates.md]: the `sbt new` command

File: src/reference/02-DetailTopics/05-Plugins-and-Best-Practices/06-sbt-new-and-Templates.md [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+to provide support for something other than Giter8 templates.
+You can skip this section if your not interested in extending `new`.
```
Comment (dwijnand) [inline src/reference/02-DetailTopics/05-Plugins-and-Best-Practices/06-sbt-new-and-Templates.md]: s/your/you're/

File: src/reference/02-DetailTopics/05-Plugins-and-Best-Practices/06-sbt-new-and-Templates.md [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+#### Template Resolver
+
+A template resolver is a partial function that looks at the arguments
```
Comment (dwijnand) [inline src/reference/02-DetailTopics/05-Plugins-and-Best-Practices/06-sbt-new-and-Templates.md]: resolving a `ModuleID`

File: src/reference/02-DetailTopics/05-Plugins-and-Best-Practices/06-sbt-new-and-Templates.md [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+after `sbt new` and determines whether it can resolve to a particular template. This is analogous to `resolvers` resolving `ModuleID` from the Internet.
+
+The `Giter8TemplateResolver` checks whether the first non-option argument looks like
+a Github repo or a git repo that ends in ".g8".
+If it matches the one of the patterns, it will pass the arguments to Giter8 to process.
```
Comment (dwijnand) [inline src/reference/02-DetailTopics/05-Plugins-and-Best-Practices/06-sbt-new-and-Templates.md]: If it matches one of the patterns

File: src/reference/02-DetailTopics/05-Plugins-and-Best-Practices/06-sbt-new-and-Templates.md [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+The rest of this page explains how to extend `sbt new` command
+to provide support for something other than Giter8 templates.
+You can skip this section if your not interested in extending `new`.
+
+#### Template Resolver
+
+A template resolver is a partial function that looks at the arguments
+after `sbt new` and determines whether it can resolve to a particular template. This is analogous to `resolvers` resolving `ModuleID` from the Internet.
+
+The `Giter8TemplateResolver` checks whether the first non-option argument looks like
```
Comment (dwijnand) [inline src/reference/02-DetailTopics/05-Plugins-and-Best-Practices/06-sbt-new-and-Templates.md]: I would remove "non-option", or at the very minimum change it to non-optional or required or positional or something - best to remove it imo

File: src/reference/02-DetailTopics/05-Plugins-and-Best-Practices/06-sbt-new-and-Templates.md [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+a Github repo or a git repo that ends in ".g8".
+If it matches the one of the patterns, it will pass the arguments to Giter8 to process.
+
+To create your own template resolver, add create a library that has `template-resolver` as a dependency:
```
Comment (dwijnand) [inline src/reference/02-DetailTopics/05-Plugins-and-Best-Practices/06-sbt-new-and-Templates.md]: Remove add

File: src/reference/02-DetailTopics/05-Plugins-and-Best-Practices/06-sbt-new-and-Templates.md [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+
+```scala
+val templateResolverApi = "org.scala-sbt" % "template-resolver" % "0.1"
+```
+
+and extend the `TemplateResolver`:
```
Comment (dwijnand) [inline src/reference/02-DetailTopics/05-Plugins-and-Best-Practices/06-sbt-new-and-Templates.md]: and extend `TemplateResolver`, which is defined as:

File: src/reference/02-DetailTopics/05-Plugins-and-Best-Practices/06-sbt-new-and-Templates.md [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+   */
+  public void run(String[] arguments);
+}
+```
+
+Publish the library to sbt community repo or Maven Central.
+
+#### templateResolverInfos
+
+Next, create an sbt plugin that adds `TemplateResolverInfo` to `templateResolverInfos`.
```
Comment (dwijnand) [inline src/reference/02-DetailTopics/05-Plugins-and-Best-Practices/06-sbt-new-and-Templates.md]: that add a `TemplateResolverInfo`

File: src/reference/02-DetailTopics/05-Plugins-and-Best-Practices/06-sbt-new-and-Templates.md [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+  override lazy val globalSettings: Seq[Setting[_]] =
+    Seq(
+      templateResolverInfos +=
+        TemplateResolverInfo(ModuleID("org.scala-sbt.sbt-giter8-resolver", "sbt-giter8-resolver", "0.1.0") cross CrossVersion.binary,
+          "sbtgiter8resolver.Giter8TemplateResolver")
+    )
+}
+```
+
+This indirecton allows template resolvers to have an isolated classpath independent from the rest of the build.
```
Comment (dwijnand) [inline src/reference/02-DetailTopics/05-Plugins-and-Best-Practices/06-sbt-new-and-Templates.md]: No need for both "isolated" and "independent". Perhaps: "... to have an classpath independent from the rest of the build"

**Comment by eed3si9n** [comment]
@dwijnand It is on a wip branch of sbt/website. I'll deal with addressing the comments :)

File: src/reference/02-DetailTopics/05-Plugins-and-Best-Practices/06-sbt-new-and-Templates.md [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+---
+
+  [giter8]: http://www.foundweekends.org/giter8/
+  [foundweekends]: http://www.foundweekends.org/
+  [CC0]: https://creativecommons.org/publicdomain/zero/1.0/
+
+sbt new and Templates
+---------------------
+
+sbt 0.13.13 adds `new` command, which helps create new build definitions.
```
Comment (eed3si9n) [inline src/reference/02-DetailTopics/05-Plugins-and-Best-Practices/06-sbt-new-and-Templates.md]: The term "project" is ambiguous with "subproject". The preferred term is build definition.

File: src/reference/02-DetailTopics/05-Plugins-and-Best-Practices/06-sbt-new-and-Templates.md [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+The `new` command is extensible via a mechanism called the template resolver.
```
Comment (eed3si9n) [inline src/reference/02-DetailTopics/05-Plugins-and-Best-Practices/06-sbt-new-and-Templates.md]: ok

File: src/reference/02-DetailTopics/05-Plugins-and-Best-Practices/06-sbt-new-and-Templates.md [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+
+### Trying new command
```
Comment (eed3si9n) [inline src/reference/02-DetailTopics/05-Plugins-and-Best-Practices/06-sbt-new-and-Templates.md]: There's a bug in Pamflet that doesn't render headers correctly if you format it.

File: src/reference/02-DetailTopics/05-Plugins-and-Best-Practices/06-sbt-new-and-Templates.md (hunk 3 above)
Comment (eed3si9n) [inline src/reference/02-DetailTopics/05-Plugins-and-Best-Practices/06-sbt-new-and-Templates.md]: Maybe for the readme lingo, but I think putting header on each source file adds unnecessary noise / ambiguity if it survives the template application.

File: src/reference/02-DetailTopics/05-Plugins-and-Best-Practices/06-sbt-new-and-Templates.md (hunk 8 above)
Comment (eed3si9n) [inline src/reference/02-DetailTopics/05-Plugins-and-Best-Practices/06-sbt-new-and-Templates.md]: Here's what I mean by non-option:

The `Giter8TemplateResolver` takes the first argument that does not start with a hyphen (`-`), and checks whether it looks like
a Github repo or a git repo that ends in ".g8".

File: src/reference/02-DetailTopics/05-Plugins-and-Best-Practices/06-sbt-new-and-Templates.md [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+
+To get started, you need to first install sbt 0.13.13 launcher or above.
```
Comment (eed3si9n) [inline src/reference/02-DetailTopics/05-Plugins-and-Best-Practices/06-sbt-new-and-Templates.md]: I think that should be covered somewhere else, like getting started guide's setup section.

**Pull request merged**